### PR TITLE
perf(build): use `jekyll-include-cache` plugin to reduce build time

### LINKED ISSUE #1094: Caching of expensive templates during rendering (trending-tags)
Rendering `_includes/trending-tags.html` for site with many tags / posts is quite slow.

## Description

In an unmodified new repo from the chirpy template, as well as on a fresh checkout of main. Importing a blog with 330 posts including 750 tags the site build takes about 4 minutes timed with `jekyll s --profile`

```
Build Process Summary:

| PHASE    |     TIME |
+----------+----------+
| RESET    |   0.0001 |
| READ     |   3.2577 |
| GENERATE |   0.0311 |
| RENDER   | 238.4877 |
| CLEANUP  |   0.0143 |
| WRITE    |   0.1458 |

Site Render Stats:

| Filename                                                           | Count |      Bytes |    Time |
+--------------------------------------------------------------------+-------+------------+---------+
| _includes/trending-tags.html                                       |  2346 | 225841.45K | 227.050 |
| _layouts/page.html                                                 |  1173 | 122030.48K | 121.067 |
| _layouts/default.html                                              |  1173 | 252854.99K | 116.072 |
| _includes/search-results.html                                      |  1173 | 113272.39K | 113.785 |
| _includes/related-posts.html                                       |   327 |   3533.14K |   6.505 |
| _includes/head.html                                                |  1173 |   8299.20K |   1.181 |

...
```

## To Reproduce

Create a new chirpy site from the template and place many posts that have tags into the _posts directory.

## Desired behavior

It would be nice if templates that are known to produce the same output site-wide could somehow be cached while rendering.

I experimented with the [jekyll-include-cache](https://github.com/benbalter/jekyll-include-cache) and saw good results.

By adding `jekyll-include-cache` as a plugin and then overriding the _layout & _includes files which reference trending-tags.html I was able to render the same site in ~11 seconds.

```
Build Process Summary:

| PHASE    |    TIME |
+----------+---------+
| RESET    |  0.0001 |
| READ     |  1.6198 |
| GENERATE |  0.0346 |
| RENDER   | 10.6490 |
| CLEANUP  |  0.0087 |
| WRITE    |  0.1347 |


Site Render Stats:

| Filename                                                           | Count |      Bytes |  Time |
+--------------------------------------------------------------------+-------+------------+-------+
| _layouts/page.html                                                 |  1172 | 121722.75K | 7.679 |
| jekyll-theme-chirpy-6.0.1/_includes/related-posts.html             |   327 |   3533.11K | 6.397 |
| _layouts/default.html                                              |  1172 | 248347.79K | 1.818 |
| jekyll-theme-chirpy-6.0.1/_includes/update-list.html               |  1172 |     72.11K | 0.897 |

...
```

### diff generated with this git command
```shell
git diff --no-index --diff-filter=M $(bundle info --path jekyll-theme-chirpy)/_layouts _layouts
git diff --no-index --diff-filter=M $(bundle info --path jekyll-theme-chirpy)/_includes _includes
```
  
```diff
diff --git a/Users/jakerr/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/jekyll-theme-chirpy-6.0.1/_layouts/page.html b/./_layouts/page.html
index 447977d..233ee18 100644
--- a/Users/jakerr/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/jekyll-theme-chirpy-6.0.1/_layouts/page.html
+++ b/./_layouts/page.html
@@ -45,7 +45,7 @@ layout: default
   <div id="panel-wrapper" class="col-xl-3 ps-2 text-muted">
     <div class="access">
       {% include update-list.html lang=lang %}
-      {% include trending-tags.html lang=lang %}
+      {% include_cached trending-tags.html lang=lang %}
     </div>
 
     {% for _include in layout.panel_includes %}
```

```diff
diff --git a/Users/jakerr/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/jekyll-theme-chirpy-6.0.1/_includes/search-results.html b/_includes/search-results.html
index e4ffb48..07981ff 100644
--- a/Users/jakerr/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/jekyll-theme-chirpy-6.0.1/_includes/search-results.html
+++ b/_includes/search-results.html
@@ -3,7 +3,7 @@
 <div id="search-result-wrapper" class="d-flex justify-content-center unloaded">
   <div class="col-11 post-content">
     <div id="search-hints">
-      {% include trending-tags.html %}
+      {% include_cached trending-tags.html %}
     </div>
     <div id="search-results" class="d-flex flex-wrap justify-content-center text-muted mt-3"></div>
   </div>

```

## Environment

| Command                           | Version |
|-----------------------------------|---------|
| `ruby -v`                         | ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22] |
| `gem -v`                          | 3.4.14 |
| `bundle -v`                       | Bundler version 2.4.14 |
| `bundle exec jekyll -v`           | jekyll 4.3.2 |
| `bundle info jekyll-theme-chirpy` | jekyll-theme-chirpy (6.0.1) |

### Desktop

- OS: [e.g. macOS 13.4]

### PR DESCRIPTION
**Pull request by cotes2020**
## Description

Resolves #1094

## Type of change

- [x] Improvement (refactoring and improving code)

## Additional context

Improvements to: 
- #1097

## How has this been tested

- [x] I have run `bash ./tools/test` (at the root of the project) locally and passed
- [x] I have tested this feature in the browser

### CODE REVIEW

**Comment by cotes2020** [comment]
Hi @jakerr, the include of the dynamic file is removed in this PR, what do you think?

File: _layouts/default.html [hunk 1]
```diff
@@ -20,17 +20,17 @@
   {% include head.html %}
 
   <body>
-    {% include sidebar.html lang=lang %}
+    {% include_cached sidebar.html lang=lang %}
```
Comment (jakerr) [inline _layouts/default.html:23]: I think you should revert this one. The `sidebar.html` template depends on the page information to decide whether to highlight with `active` css class.

So the sidebar will still work without it but this highlight would disappear:
Screenshot 2023-06-19 at 11 51 29

Luckily this template is not to expensive in a large site anyhow:
eg:
```
| Filename                                                           | Count |      Bytes |  Time |
+--------------------------------------------------------------------+-------+------------+-------+
| jekyll-theme-chirpy-6.0.1/_includes/sidebar.html                   |  1168 |   2961.32K | 0.299 |
```

File: _layouts/tags.html [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # All the Tags of posts.
 ---
 
-<div id="tags" class="d-flex flex-wrap mx-xl-2">
+<div id="tags" class="d-flex flex-wrap mx-xl-2 mb-5">
```
Comment (jakerr) [inline _layouts/tags.html:6]: Was this intentional?

**Review by jakerr — COMMENTED** [review]
Looks great thank you!
I just found the one case where the template needs dynamic page info (the sidebar) I checked all of the other templates and they look good to cache to me.

Here's the perf for this change checked locally (on this site it was 4 minutes without this change):

```
Build Process Summary: 

| PHASE    |   TIME |
+----------+--------+
| RESET    | 0.0001 |
| READ     | 2.0448 |
| GENERATE | 0.0375 |
| RENDER   | 9.8069 |
| CLEANUP  | 0.0042 |
| WRITE    | 0.1675 |
 

Site Render Stats: 

| Filename                                                           | Count |      Bytes |  Time |
+--------------------------------------------------------------------+-------+------------+-------+
| _layouts/page.html                                                 |  1168 | 122040.89K | 6.781 |
| jekyll-theme-chirpy-6.0.1/_includes/related-posts.html             |   326 |   3514.48K | 6.367 |
| _layouts/default.html                                              |  1168 | 248613.28K | 1.806 |
| _includes/head.html                                                |  1168 |   5208.82K | 0.810 |
| jekyll-theme-chirpy-6.0.1/_includes/js-selector.html               |  1168 |    854.92K | 0.300 |
| jekyll-theme-chirpy-6.0.1/_includes/sidebar.html                   |  1168 |   2961.32K | 0.299 |
```

**Comment by jakerr** [comment]
Oh one more thing. How are you enabling the plugin? Is it automatic?
I thought I had to add this in _config.yml:

```
plugins:
  - jekyll-include-cache
```

File: _layouts/tags.html (hunk 2 above)
Comment (cotes2020) [reply _layouts/tags.html:6]: Yes, when the number of tags is too large and exceeds the height of the screen, a margin needs to be added at the bottom.

Example for 1000 tags:

Screen Shot 2023-06-19 at 13 13 30

File: _layouts/default.html (hunk 1 above)
Comment (cotes2020) [reply _layouts/default.html:23]: Good catch!

**Comment by cotes2020** [comment]
> Oh one more thing. How are you enabling the plugin? Is it automatic? I thought I had to add this in _config.yml:
> 
> ```
> plugins:
>   - jekyll-include-cache
> ```

For the source code project here, the plugin is referenced in the `jekyll-theme-chirpy.gemspec` file.

```ruby
spec.add_runtime_dependency "jekyll-include-cache", "~> 0.2"
```

For users who use the gem package, currently until the new [Chirpy gem](https://rubygems.org/gems/jekyll-theme-chirpy) is released, the `jekyll-include-cache` plugin needs to be imported in the `Gemfile`.

**Comment by jakerr** [comment]
> For the source code project here, the plugin is referenced in the `jekyll-theme-chirpy.gemspec` file.
> 

Ah thanks. I wasn't aware of the difference between depending on _plugins in a Jekyll site and depending on them from normal libs when writing plugins. [Found a bit of info on that here](https://talk.jekyllrb.com/t/plugins-and-authored-theme-gems/7595).

This PR looks good to me.

**Comment by cotes2020** [comment]
In fact, there is no need to modify `_config.yml` to use the plugin, you can use [the `:jekyll_plugins` Gemfile group](https://jekyllrb.com/docs/plugins/installation/#the-jekyll_plugins-group)

**Pull request merged**